agent/src/agentclient.rs

```diff
@@ -20,90 +25,139 @@ use models::{
     },
 };
 
+// The reflector uses exponential backoff.
+// These values configure how long to delay between tries.
+const RETRY_BASE_DELAY: Duration = Duration::from_millis(1000);
+const RETRY_MAX_DELAY: Duration = Duration::from_secs(10);
+const NUM_RETRIES: usize = 5;
+
 const AGENT_SLEEP_DURATION: Duration = Duration::from_secs(5);
 
 #[derive(Clone)]
 pub struct BrupopAgent<T: APIServerClient> {
     k8s_client: kube::client::Client,
-    node_selector: Option<BottlerocketNodeSelector>,
     apiserver_client: T,
+    brn_reader: Store<BottlerocketNode>,
+    node_reader: Store<Node>,
+    associated_node_name: String,
+    associated_bottlerocketnode_name: String,
 }
 
 impl<T: APIServerClient> BrupopAgent<T> {
-    pub fn new(k8s_client: kube::client::Client, apiserver_client: T) -> Self {
+    pub fn new(
+        k8s_client: kube::client::Client,
+        apiserver_client: T,
+        brn_reader: Store<BottlerocketNode>,
+        node_reader: Store<Node>,
+        associated_node_name: String,
+        associated_bottlerocketnode_name: String,
+    ) -> Self {
         BrupopAgent {
             k8s_client,
             apiserver_client,
-            node_selector: None,
+            brn_reader,
+            node_reader,
+            associated_node_name,
+            associated_bottlerocketnode_name,
         }
     }
 
     pub async fn check_node_custom_resource_exists(&mut self) -> Result<bool> {
-        let associated_bottlerocketnode_name = self.get_node_selector().await?.brn_resource_name();
-        let bottlerocket_nodes: Api<BottlerocketNode> =
-            Api::namespaced(self.k8s_client.clone(), NAMESPACE);
-
-        Ok(bottlerocket_nodes
-            .get(&associated_bottlerocketnode_name)
-            .await
-            .context(error::BottlerocketNodeNotExist {
-                node_name: associated_bottlerocketnode_name,
-            })
-            .is_ok())
+        // try to check if node custom resource exist in the store first. If it's not present in the
+        // store(either node custom resource doesn't exist or store data delays), make the API call for second round check.
+
+        let associated_bottlerocketnode = self.brn_reader.state().clone();
+        if associated_bottlerocketnode.len() != 0 {
+            Ok(true)
+        } else {
+            let bottlerocket_nodes: Api<BottlerocketNode> =
+                Api::namespaced(self.k8s_client.clone(), NAMESPACE);
+
+            // handle the special case which custom resource does exist but communication with the k8s API fails for other errors.
+            if let Err(e) = bottlerocket_nodes
+                .get(&self.associated_bottlerocketnode_name.clone())
+                .await
```

Comment:
Is this fetch from the reflector first and then try to get from k8s etcd? Why do you need fetch from k8s etcd again? Is it because reflector is like cache that you need miss once before the data get moved inside, in this case why does `fetch_custom_resource` only fetch from reflector and error out if failed?

Reply:
No, We tried to fetch from the reflector first. If it doesn't work because of delay on reflector . Sometime when I query `Store`, it return empty to me, which would be recognized BRN had not been created, but it just because of delay. Therefore, If delay on reflector, we will fetch k8s etcd to get the metadata. Here is the discuss with Sean https://github.com/bottlerocket-os/bottlerocket-update-operator/pull/118#discussion_r750732385